### PR TITLE
 [BEAM-3490] Make runtime type checking code runner agnostic.

### PR DESCRIPTION
**Pull request by robertwb**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [ ] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

**Comment by robertwb** [comment]
R: @charlesccychen

File: sdks/python/apache_beam/typehints/typecheck.py [hunk 1]
```diff
@@ -178,3 +180,74 @@ def _type_check(self, type_constraint, datum, is_input):
                    "Instead, received '%s', an instance of type %s."
                    % (datum_type, type_constraint, datum, type(datum)))
       raise TypeCheckError, error_msg, sys.exc_info()[2]
+
+
+class TypeCheckCombineFn(core.CombineFn):
+  """A wrapper around a DoFn which performs type-checking of input and output.
```
Comment (charlesccychen) [inline sdks/python/apache_beam/typehints/typecheck.py]: Please fix doc string: `CombineFn`.

File: sdks/python/apache_beam/typehints/typecheck.py [hunk 2]
```diff
@@ -178,3 +180,74 @@ def _type_check(self, type_constraint, datum, is_input):
...
+
+  def create_accumulator(self, *args, **kwargs):
+    print 'create_accumulator', self._combinefn, args, kwargs
+    return self._combinefn.create_accumulator(*args, **kwargs)
+
+  def add_input(self, accumulator, element, *args, **kwargs):
+    if self._input_type_hint:
+      try:
+        _check_instance_type(
+            self._input_type_hint[0][0].tuple_types[1], element, 'element', True)
```
Comment (charlesccychen) [inline sdks/python/apache_beam/typehints/typecheck.py]: Please fix line length.

File: sdks/python/apache_beam/typehints/typecheck.py [hunk 3]
```diff
@@ -35,6 +37,7 @@
 from apache_beam.typehints.typehints import CompositeTypeHintError
 from apache_beam.typehints.typehints import SimpleTypeHintError
 from apache_beam.typehints.typehints import check_constraint
```
Comment (charlesccychen) [inline sdks/python/apache_beam/typehints/typecheck.py]: Please remove.

File: sdks/python/apache_beam/typehints/typecheck.py [hunk 4]
```diff
@@ -178,3 +180,74 @@ def _type_check(self, type_constraint, datum, is_input):
...
+
+  _in_combine = False
+
+  def enter_composite_transform(self, applied_transform):
+    if isinstance(applied_transform.transform, core.CombinePerKey):
+      self._in_combine = True
+      self._wrapped_fn = applied_transform.transform.fn = TypeCheckCombineFn(
+          applied_transform.transform.fn,
+          applied_transform.transform.get_type_hints(),
+          applied_transform.full_label)
```
Comment (charlesccychen) [inline sdks/python/apache_beam/typehints/typecheck.py]: Please add newline.

File: sdks/python/apache_beam/typehints/typecheck.py [hunk 5]
```diff
@@ -178,3 +180,74 @@ def _type_check(self, type_constraint, datum, is_input):
...
+  """
+
+  def __init__(self, combinefn, type_hints, label=None):
+    self._combinefn = combinefn
+    self._input_type_hint = type_hints.input_types
+    self._output_type_hint = type_hints.simple_output_type(label)
+    self._label = label
+
+  def create_accumulator(self, *args, **kwargs):
+    print 'create_accumulator', self._combinefn, args, kwargs
```
Comment (charlesccychen) [inline sdks/python/apache_beam/typehints/typecheck.py]: Please remove debugging output.

File: sdks/python/apache_beam/runners/portability/fn_api_runner.py [hunk 6]
```diff
@@ -192,6 +194,9 @@ def _next_uid(self):
 
   def run_pipeline(self, pipeline):
     MetricsEnvironment.set_metrics_supported(False)
+    # This is sometimes needed if type checking is disabled.
```
Comment (charlesccychen) [inline sdks/python/apache_beam/runners/portability/fn_api_runner.py]: Can you clarify in more detail what this does?

File: sdks/python/apache_beam/typehints/typecheck.py (hunk 1 above)
Comment (robertwb) [inline sdks/python/apache_beam/typehints/typecheck.py]: > **charlesccychen** wrote:
> Please fix doc string: `CombineFn`.


Done.

File: sdks/python/apache_beam/typehints/typecheck.py (hunk 2 above)
Comment (robertwb) [inline sdks/python/apache_beam/typehints/typecheck.py]: > **charlesccychen** wrote:
> Please fix line length.


Done.

File: sdks/python/apache_beam/typehints/typecheck.py (hunk 4 above)
Comment (robertwb) [inline sdks/python/apache_beam/typehints/typecheck.py]: > **charlesccychen** wrote:
> Please add newline.


Done.

File: sdks/python/apache_beam/typehints/typecheck.py (hunk 3 above)
Comment (robertwb) [inline sdks/python/apache_beam/typehints/typecheck.py]: > **charlesccychen** wrote:
> Please remove.


Done.

File: sdks/python/apache_beam/typehints/typecheck.py (hunk 5 above)
Comment (robertwb) [inline sdks/python/apache_beam/typehints/typecheck.py]: > **charlesccychen** wrote:
> Please remove debugging output.


Done.

File: sdks/python/apache_beam/runners/portability/fn_api_runner.py (hunk 6 above)
Comment (robertwb) [inline sdks/python/apache_beam/runners/portability/fn_api_runner.py]: > **charlesccychen** wrote:
> Can you clarify in more detail what this does?


Done.

**Pull request merged**